src/util/syscall_sandbox.cpp

```diff
@@ -0,0 +1,488 @@
+    assert(SYSCALL_NAMES.size() == 333 && SYSCALL_NAMES[0] == "read" && SYSCALL_NAMES[332] == "statx" && "Syscalls must be listed in syscall number order without gaps.");
+    return syscall_number < SYSCALL_NAMES.size() ? SYSCALL_NAMES[syscall_number] : "*unknown*";
+}
+
+void SyscallSandboxDebugSignalHandler(int, siginfo_t* info, void* void_context)
+{
+    if (info->si_code != SYS_SECCOMP) {
+        return;
+    }
+    const ucontext_t* ctx = static_cast<ucontext_t*>(void_context);
+    if (ctx == nullptr) {
+        return;
+    }
+    std::set_new_handler(std::terminate);
+    const unsigned int syscall_number = ctx->uc_mcontext.gregs[REG_SYSCALL];
+    const std::string syscall_name = GetLinuxSyscallName(syscall_number);
+    const std::string thread_name = !util::ThreadGetInternalName().empty() ? util::ThreadGetInternalName() : "*unnamed*";
+    const std::string error_message = strprintf("ERROR: The syscall \"%s\" (syscall number %d) is not allowed by the syscall sandbox in thread \"%s\". Please report. Exiting.", syscall_name, syscall_number, thread_name);
+    tfm::format(std::cerr, "%s\n", error_message);
+    LogPrintf("%s\n", error_message);
+    std::terminate();
+}
+
+void InstallSyscallSandboxDebugHandler()
+{
+    static std::atomic<bool> syscall_reporter_installed{false};
+    if (syscall_reporter_installed.exchange(true)) {
+        return;
+    }
+    LogPrint(BCLog::UTIL, "Installing syscall sandbox debug handler\n");
+    struct sigaction action = {};
+    sigset_t mask;
+    sigemptyset(&mask);
+    sigaddset(&mask, SIGSYS);
+    action.sa_sigaction = &SyscallSandboxDebugSignalHandler;
+    action.sa_flags = SA_SIGINFO;
+    if (sigaction(SIGSYS, &action, nullptr) < 0) {
+        perror("sigaction");
```

Comment:
In commit "Add syscall sandboxing (seccomp-bpf)" (3636a5c1f03fd01da2a78cad0cba7ab69cadf074)

Throughout this file is there a reason error handling code can't use normal InitError or LogPrintf functions or raise exceptions, to send errors to debug.log and the GUI instead of bypassing them and going to stderr?

Reply:
Good point. The initialization/setup logic has now been moved to `init.cpp` and it is now using the expected `InitError` and `LogPrintf`. `SetSyscallSandboxPolicy` is now throwing `std::runtime_error` in case of errors.